### PR TITLE
Add support for replication of object tags, retention metadata

### PR DESCRIPTION
**Pull request by poornas**
## Motivation and Context
replication of object tags and retention/legal hold metdata currently missing functionality

## How to test this PR?
`mc cp` on a replication enabled bucket, then `mc tag set` on a version or `mc retention`

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

### CODE REVIEW

**Comment by minio-trusted** [comment]
```
Error running mint automation
+ mkdir -p 10880-7fab924/gopath/src/github.com/minio
+ git -C 10880-7fab924/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 10880-7fab924/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 10880-7fab924/gopath/src/github.com/minio/minio fetch --quiet minio pull/10880/head:pr10880
+ git -C 10880-7fab924/gopath/src/github.com/minio/minio checkout --quiet pr10880
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/10880-7fab924/gopath
+ make -C 10880-7fab924/gopath/src/github.com/minio/minio --quiet
go: github.com/minio/minio-go/v7@v7.0.6-0.20201013215222-14baba9e61ac: parsing ../minio-go/go.mod: open /home/mint-auto/mint-auto/10880-7fab924/gopath/src/github.com/minio/minio-go/go.mod: no such file or directory
go: github.com/minio/minio-go/v7@v7.0.6-0.20201013215222-14baba9e61ac: parsing ../minio-go/go.mod: open /home/mint-auto/mint-auto/10880-7fab924/gopath/src/github.com/minio/minio-go/go.mod: no such file or directory
make: *** [build] Error 1

```

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#6e0a82d7-2142-401a-9037-f502d71ba2f8) |
<a name="6e0a82d7-2142-401a-9037-f502d71ba2f8"></a>
### 10880-00faaa9/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:31549
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 885c77382459:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 21 seconds
(4/15) Running aws-sdk-ruby tests ... done in 21 seconds
(5/15) Running awscli tests ... done in 3 minutes and 7 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 28 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 39 seconds
(9/15) Running minio-go tests ... done in 6 minutes and 24 seconds
(10/15) Running minio-java tests ... done in 10 minutes and 44 seconds
(11/15) Running minio-js tests ... done in 2 minutes and 55 seconds
(12/15) Running minio-py tests ... done in 18 minutes and 41 seconds
(13/15) Running s3cmd tests ... done in 2 minutes and 16 seconds
(14/15) Running s3select tests ... FAILED in 11 seconds
{
  "name": "s3select:test_csv_input_quote_char",
  "function": "select_object_content(bucket_name, object_name, opts)",
  "args": {
    "bucket_name": "s3select-test-f33c5854-5215-4f23-a484-4d754df1b12c"
  },
  "duration": 11544,
  "message": "Test test_13 unexpectedly failed with: 'event-type'",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/s3select/csv.py\", line 39, in test_sql_api\n    for d in data.stream(10*1024):\n  File \"/usr/local/lib/python3.6/dist-packages/minio/select/reader.py\", line 211, in stream\n    message = self.__extract_message()\n  File \"/usr/local/lib/python3.6/dist-packages/minio/select/reader.py\", line 161, in __extract_message\n    event_type = header_map[\"event-type\"]\nKeyError: 'event-type'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"./tests.py\", line 52, in main\n    test_csv_input_custom_quote_char(client, log_output)\n  File \"/mint/run/core/s3select/csv.py\", line 119, in test_csv_input_custom_quote_char\n    data, sql_opts, expected_output)\n  File \"/mint/run/core/s3select/csv.py\", line 45, in test_sql_api\n    'Test {} unexpectedly failed with: {}'.format(test_name, select_err))\nValueError: Test test_13 unexpectedly failed with: 'event-type'\n",
  "status": "FAIL"
}
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by cesnietor — APPROVED** [review]
Tested adding/removing tags/legalhold works well.

File: cmd/data-crawler.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ import (
 const (
 	dataCrawlSleepPerFolder  = time.Millisecond // Time to wait between folders.
 	dataCrawlSleepDefMult    = 10.0             // Default multiplier for waits between operations.
-	dataCrawlStartDelay      = 5 * time.Minute  // Time to wait on startup and between cycles.
+	dataCrawlStartDelay      = 1 * time.Minute  // Time to wait on startup and between cycles.
```
Comment (harshavardhana) [inline cmd/data-crawler.go]:
```suggestion
	dataCrawlStartDelay      = 5 * time.Minute  // Time to wait on startup and between cycles.
```

File: cmd/erasure-object.go [hunk 2]
```diff
@@ -1096,6 +1099,11 @@ func (er erasureObjects) PutObjectTags(ctx context.Context, bucket, object strin
 		if tags != "" {
 			fi.Metadata[xhttp.AmzObjectTagging] = tags
 		}
+		if opts.UserDefined != nil {
```
Comment (harshavardhana) [inline cmd/erasure-object.go]: not needed just the following line is enough.

```suggestion
		for k, v := range opts.UserDefined {
		}
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/bucket-replication.go [hunk 3]
```diff
@@ -241,6 +242,45 @@ func replicateDelete(ctx context.Context, dobj DeletedObjectVersionInfo, objectA
 	}
 }
 
+func getCopyObjMetadata(oi ObjectInfo, dest replication.Destination) map[string]string {
+	meta := make(map[string]string)
```
Comment (harshavardhana) [inline cmd/bucket-replication.go]:
```suggestion
	meta := make(map[string]string, len(oi.UserDefined))
```

File: cmd/bucket-replication.go [hunk 4]
```diff
@@ -294,6 +334,53 @@ func putReplicationOpts(ctx context.Context, dest replication.Destination, objIn
...
+type replicationType string
+
+const (
+	replicateMetadata replicationType = "metadata"
+	replicateNone     replicationType = "none"
+	replicateAll      replicationType = "all"
+)
+
+// returns replicationType by comparing metadata between source and target
+func getReplicationType(oi1 ObjectInfo, oi2 minio.ObjectInfo) replicationType {
```
Comment (harshavardhana) [inline cmd/bucket-replication.go]:
```suggestion
func getReplicateAction(oi1 ObjectInfo, oi2 minio.ObjectInfo) replicateAction {
```

File: cmd/bucket-replication.go [hunk 5]
```diff
@@ -294,6 +334,53 @@ func putReplicationOpts(ctx context.Context, dest replication.Destination, objIn
 	return
 }
 
+type replicationType string
+
+const (
+	replicateMetadata replicationType = "metadata"
+	replicateNone     replicationType = "none"
+	replicateAll      replicationType = "all"
+)
```
Comment (harshavardhana) [inline cmd/bucket-replication.go]:
```suggestion
type replicateAction string

const (
	replicateMetadata replicateAction = "metadata"
	replicateNone     replicateAction = "none"
	replicateAll      replicateAction = "all"
)
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#d6074bfc-478d-4d9b-b0b4-f7453c9fbc4f) |
<a name="d6074bfc-478d-4d9b-b0b4-f7453c9fbc4f"></a>
### 10880-5daacb7/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:30185
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6704d39ba5fe:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 26 seconds
(4/15) Running aws-sdk-ruby tests ... done in 20 seconds
(5/15) Running awscli tests ... done in 2 minutes and 54 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 32 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 37 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 50 seconds
{
  "args": {
    "bucketName": "minio-go-test-tgx14hf3f0ir5t6r",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 1176,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 11 minutes and 20 seconds
(10/15) Running minio-js tests ... FAILED in 46 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 79,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:682:10) at endReadableNT (_stream_readable.js:1252:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... done in 19 minutes and 13 seconds
(11/15) Running s3cmd tests ... done in 2 minutes and 24 seconds
(12/15) Running s3select tests ... done in 40 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**